src/django_components/util/testing.py

```diff
@@ -0,0 +1,497 @@
+import inspect
+import os
+import sys
+from functools import wraps
+from typing import Any, Callable, Dict, Iterable, List, Optional, Sequence, Set, Tuple, Type, Union
+from unittest.mock import patch
+from weakref import ReferenceType
+
+from django.template import engines
+from django.conf import settings as _django_settings
+from django.test import override_settings
+
+from django_components.autodiscovery import LOADED_MODULES
+from django_components.component import ALL_COMPONENTS, Component, component_node_subclasses_by_name
+from django_components.component_media import ComponentMedia
+from django_components.component_registry import ALL_REGISTRIES, ComponentRegistry
+
+
+class GenIdPatcher:
+    def __init__(self):
+        self._gen_id_count = 10599485
+        self._gen_id_patch = None
+
+    # Mock the `generate` function used inside `gen_id` so it returns deterministic IDs
+    def start(self):
+        # Random number so that the generated IDs are "hex-looking", e.g. a1bc3d
+        self._gen_id_count = 10599485
+
+        def mock_gen_id(*args, **kwargs):
+            self._gen_id_count += 1
+            return hex(self._gen_id_count)[2:]
+
+        self._gen_id_patch = patch("django_components.util.misc.generate", side_effect=mock_gen_id)
+        self._gen_id_patch.start()
+
+    def stop(self):
+        if not self._gen_id_patch:
+            return
+
+        self._gen_id_patch.stop()
+        self._gen_id_patch = None
+
+
+class CsrfTokenPatcher:
+    def __init__(self):
+        self._csrf_token = "predictabletoken"
+        self._csrf_token_patch = None
+
+    def start(self):
+        self._csrf_token_patch = patch("django.middleware.csrf.get_token", return_value=self._csrf_token)
+        self._csrf_token_patch.start()
+
+    def stop(self):
+        if self._csrf_token_patch:
+            self._csrf_token_patch.stop()
+            self._csrf_token_patch = None
+
+
+def djc_test(
```

Comment:
Here is the actual definition of the decorator. Much of the logic was previously in `tests/testutils.py`.